allo/backend/aie.py

```diff
@@ -221,86 +237,153 @@ def codegen_aie_mlir(mod, orig_input_args, mapping):
     for mid in range(mem_tile_size):
         code += format_str(f"%tile_mem{mid} = aie.tile({mid}, 1)")
-    assert len(mapping) == 1, "Only support 1D mapping for now"
-    pe_size = mapping[0]
+    # assert len(mapping) == 1, "Only support 1D mapping for now"
+    # TODO: maybe use name of the function to support 2D?
+    # number of function declaration except top
+    funcs = list(mod.body.operations)[:-1]
+    pe_size = len(funcs)
+    buf_name_dicts = []
     for pid in range(pe_size):
         code += format_str(f"%tile_comp{pid} = aie.tile(0, {pid + 2})")
+        buf_dict = buf_dicts[pid]
+        buf_name_dict = {}
+        for i, name in enumerate(buf_dict.keys()):
+            tile_name = f"%tile_comp{pid}"
+            new_name = f"{tile_name}_buf{i}"
+            buf_name_dict[name] = new_name
+            ele_type, shape = buf_dict[name]
+            str_list = list(map(str, shape))
+            str_list.append(ele_type)
+            buf_type = f"memref<{'x'.join(map(str, str_list))}>"
+            code += format_str(f"{new_name} = aie.buffer({tile_name}) : {buf_type}")
+        buf_name_dicts.append(buf_name_dict)
     # update module and args
-    mod_str = str(mod)
-    for i, (ele_type, shape) in enumerate(input_args):
-        orig_ele_type = f"memref<{'x'.join(map(str, shape))}x{ele_type}>"
-        shape = (shape[0] // pe_size, *shape[1:])
+    for j, (ele_type, orig_shape) in enumerate(input_args):
+        orig_ele_type = f"memref<{'x'.join(map(str, orig_shape))}x{ele_type}>"
+        # TODO: need to deal with different sizes for different funcs
+        shape = func_sizes[0][j]
         ele_type = f"memref<{'x'.join(map(str, shape))}x{ele_type}>"
-        input_args[i] = (ele_type, orig_ele_type, shape)
-        mod_str = mod_str.replace(orig_ele_type, ele_type)
+        input_args[j] = (ele_type, orig_ele_type, shape, orig_shape)
+    func_strs = list(map(str, funcs))
+    # update buffers
+    for pid in range(pe_size):
+        func_str = func_strs[pid]
+        buf_name_dict = buf_name_dicts[pid]
+        # remove memref.alloc
+        pattern_alloc = re.compile(r"^.*memref\.alloc.*\n?", re.MULTILINE)
+        func_str = re.sub(pattern_alloc, "", func_str)
+        # replace new buffer name
+        pattern_boundary = r"(?<![\w.]){old}(?![\w.])"
+        for name, new_name in buf_name_dict.items():
+            escaped_name = re.escape(name)
+            pattern = pattern_boundary.format(old=escaped_name)
+            func_str = re.sub(pattern, new_name, func_str)
+        func_strs[pid] = func_str
     # create object fifos
     # connect each argument to a separate mem tile
-    for i, (in_type, orig_in_type, shape) in enumerate(input_args[:-1]):
-        # depth=2 means double buffer
-        code += format_str(
-            f"aie.objectfifo @in_sh{i}(%tile_shim, {{%tile_mem{i}}}, 2 : i32) : !aie.objectfifo<{orig_in_type}>"
-        )
+    linkings = [False] * len(input_args)
```

Comment:
Add comment for `linkings`

Reply:
Based on your description, a clearer name might be `dist_alloc`, which reflects that when set to True the memory is distributed among compute tiles, and when False it is replicated to each tile.